### PR TITLE
split CFL, CML, CNL firmware names

### PR DESCRIPTION
**Pull request by plbossart**
we need to split those firmwares due to signature issues
this replaces https://github.com/thesofproject/linux/pull/1278

### CODE REVIEW

**Comment by plbossart** [comment]
@xiulipan you'll need to change the CI settings, we still check for CNL firmware on CML hardware:

2019-10-29 15:01:56 UTC [REM_INFO]: Checking file: /lib/firmware/intel/sof/sof-cnl.ri
2019-10-29 15:01:56 UTC [REM_INFO]: Found file: /lib/firmware/intel/sof/sof-cnl.ri

**Comment by ranj063** [comment]
@plbossart do we also need the corresponding changes in sof-pci-dev.c to add a cml_desc?

**Comment by plbossart** [comment]
the changes are already there:

````c
static const struct sof_dev_desc cml_desc = {
	.machines		= snd_soc_acpi_intel_cnl_machines,		.machines		= snd_soc_acpi_intel_cnl_machines,
	.resindex_lpe_base	= 0,		.resindex_lpe_base	= 0,
	.resindex_pcicfg_base	= -1,		.resindex_pcicfg_base	= -1,
	.resindex_imr_base	= -1,		.resindex_imr_base	= -1,
	.irqindex_host_ipc	= -1,		.irqindex_host_ipc	= -1,
	.resindex_dma_base	= -1,		.resindex_dma_base	= -1,
	.chip_info = &cnl_chip_info,		.chip_info = &cnl_chip_info,
	.default_fw_path = "intel/sof",		.default_fw_path = "intel/sof",
	.default_tplg_path = "intel/sof-tplg",		.default_tplg_path = "intel/sof-tplg",
	.nocodec_fw_filename = "sof-cnl.ri",		.nocodec_fw_filename = "sof-cml.ri",
````

**Comment by ranj063** [comment]
> ```c
> 	.machines		= snd_soc_acpi_intel_cnl_machines,		.machines		= snd_soc_acpi_intel_cnl_machines,
> ```

@plbossart this is what Im confused about. Shouldnt this be snd_soc_acpi_intel_cml_machines with your new addition?

**Comment by plbossart** [comment]
> > ```c
> > 	.machines		= snd_soc_acpi_intel_cnl_machines,		.machines		= snd_soc_acpi_intel_cnl_machines,
> > ```
> 
> @plbossart this is what Im confused about. Shouldnt this be snd_soc_acpi_intel_cml_machines with your new addition?

ah yes, that's just wrong. Thanks for spotting this.

File: sound/soc/sof/sof-pci-dev.c [hunk 1]
```diff
@@ -128,7 +128,7 @@ static const struct sof_dev_desc cfl_desc = {
 	.chip_info = &cnl_chip_info,
 	.default_fw_path = "intel/sof",
 	.default_tplg_path = "intel/sof-tplg",
-	.nocodec_fw_filename = "sof-cnl.ri",
+	.nocodec_fw_filename = "sof-cfl.ri",
```
Comment (RanderWang) [inline sound/soc/sof/sof-pci-dev.c:132]: for cfl,   we use sof-cnl.ri for HDA platform, why we need to use sof-cfl.ri for nocodec ?
I think we don't need to change here.

**Comment by RanderWang** [comment]
@plbossart ,   do we need to split cml from cnl ?  please check our sdw table https://github.com/thesofproject/linux/blob/integration/soundwire-latest/sound/soc/intel/common/soc-acpi-intel-cnl-match.c#L68.  we could just split cnl table to cnl & cml table, and I still consider cml as a cnl platform

**Comment by xiulipan** [comment]
@plbossart Thanks for the note. We will try to come up with a smooth transition method.

File: sound/soc/sof/sof-pci-dev.c (hunk 1 above)
Comment (plbossart) [reply sound/soc/sof/sof-pci-dev.c:132]: if we start using a different firmware for all platforms, why would we leave CFL and CNL the same? That'd be the exception, so let's clean this up.

**Comment by plbossart** [comment]
@fredoh9 the CI tests should break on CFL, will need your help to add sof-cnl.ri -> sof-cfl.ri symlink

File: sound/soc/intel/common/soc-acpi-intel-cfl-match.c [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+ *
+ */
+
+#include <sound/soc-acpi.h>
+#include <sound/soc-acpi-intel-match.h>
+
+struct snd_soc_acpi_mach snd_soc_acpi_intel_cfl_machines[] = {
+	{},
+};
+EXPORT_SYMBOL_GPL(snd_soc_acpi_intel_cfl_machines);
```
Comment (kv2019i) [inline sound/soc/intel/common/soc-acpi-intel-cfl-match.c:15]: Putting an empty table looks a bit odd, but given there will be entryes, ++votes still.

**Comment by plbossart** [comment]
@fredoh9 can you help recheck the CFL and CML HDA cases. I see messages such as 

````
[    9.302490 &lt;    0.226217&gt;] sof-audio-pci 0000:00:1f.3: Direct firmware load for intel/sof/(null) failed with error -2
[    9.302497 &lt;    0.000007&gt;] sof-audio-pci 0000:00:1f.3: error: request firmware intel/sof/(null) failed err: -2
[    9.302589 &lt;    0.000092&gt;] sof-audio-pci 0000:00:1f.3: error: failed to load DSP firmware -2
[    9.302656 &lt;    0.000067&gt;] sof-audio-pci 0000:00:1f.3: error: sof_probe_work failed err: -2
````

Not sure if my fallback worked here.

File: sound/soc/sof/intel/hda.c [hunk 3]
```diff
@@ -410,9 +410,16 @@ static int hda_init_caps(struct snd_sof_dev *sdev)
...
 				hda_mach->sof_tplg_filename;
 
-			/* firmware: pick the first in machine list */
+			/*
+			 * firmware: pick the first in machine list,
+			 * or use nocodec
+			 */
 			mach = pdata->desc->machines;
-			pdata->fw_filename = mach->sof_fw_filename;
+			if (mach)
```
Comment (ranj063) [inline sound/soc/sof/intel/hda.c]: @plbossart Im not sure an empty struct is equivalent to NULL. Maybe check for ARRAY_SIZE() is a better option?

Comment (plbossart) [reply sound/soc/sof/intel/hda.c]: yes, that was my fear when I wrote the code, but then we also test for the NULL case when we reach the last empty structure in the ACPI table, when we scan for possible machines. That's why I'd want someone to test on CFL w/ HDA.

Comment (ranj063) [reply sound/soc/sof/intel/hda.c]: @plbossart one way to make sure that this is null is not to set the "machines" field until we have entries in the cfl table.
But in my multi-client series, I fixed this problem by adding a default_fw_filename field to desc because the firmware is loaded long before the acpi match is done by the audio client. But thats for another day.

Comment (plbossart) [reply sound/soc/sof/intel/hda.c]: right, and for now I am stuck due to this other issue:
https://github.com/thesofproject/linux/issues/1417

whoever invented signed firmware should buy free beer for life to all developers.

Comment (plbossart) [reply sound/soc/sof/intel/hda.c]: if (mach->id[0]) is what we used for acpi scans, so fixed the code to use the same thing

**Comment by plbossart** [comment]
@fredoh9 @xiulipan CFL and CML HDA failed with firmware name changes.

**Comment by RanderWang** [comment]
@plbossart    I got a CFL device and test this PR.  " if (mach)" always return true, and " if (mach->id[0])" works.  How about change it to
```C
			if (mach->sof_fw_filename)
				pdata->fw_filename = mach->sof_fw_filename;
			else
				pdata->fw_filename =
					pdata->desc->nocodec_fw_filename;
```

**Comment by xiulipan** [comment]
@plbossart We are working on this. I hope we do not get some sudden change to broke previous version.
But if we find we can not give the support before next weed, we may get some hard switch.

**Comment by plbossart** [comment]
> @plbossart I got a CFL device and test this PR. " if (mach)" always return true, and " if (mach->id[0])" works. How about change it to
> 
> ```c
> 			if (mach->sof_fw_filename)
> 				pdata->fw_filename = mach->sof_fw_filename;
> 			else
> 				pdata->fw_filename =
> 					pdata->desc->nocodec_fw_filename;
> ```

I guess that would probably work. I still prefer mach->id[0] for consistency with the code in soc-acpi.h
